notebook/services/kernels/handlers.py

```diff
@@ -244,6 +264,85 @@ def on_message(self, msg):
             return
         stream = self.channels[channel]
         self.session.send(stream, msg)
+        
+    def _on_zmq_reply(self, stream, msg_list):
+        
+        def write_stderr(error_message):
+            parent = json.loads(msg_list[-3])
+            msg = self.session.msg("stream",
+                content={"text": error_message, "name": "stderr"},
+                parent=parent
+            )
+            msg['channel'] = 'iopub'
+            self.write_message(json.dumps(msg, default=date_default))
+            self.log.warn(error_message)
+            
+        channel = getattr(stream, 'channel', None)
+        if channel == 'iopub':
+    
+            # If the elapsed time is double the check interval, reset the counts,
+            # lock flags, and reset the timer.
+            elapsed_time = (time.time() - self._last_limit_check)
+            if elapsed_time >= self.limit_poll_interval * 2.0:
+                self._iopub_msg_count = 0
+                self._iopub_byte_count = 0
+                if self._iopub_msgs_exceeded or self._iopub_data_exceeded:
+                    self._iopub_data_exceeded = False
+                    self._iopub_msgs_exceeded = False
+                    self.log.warn("iopub messages resumed")
+                
+            # Increment the bytes and message count
+            self._iopub_msg_count += 1
+            self._iopub_byte_count += sum([len(x) for x in msg_list])
+            
+            # If the elapsed time is equal to or greater than the limiter
+            # interval, check the limits, set the limit flags, and reset the
+            # message and data counts.
+            if elapsed_time >= self.limit_poll_interval:
```

Comment:
If poll interval is 1s, this means that I can exceed the rate limit, as long as I do it before then. There should probably be an additional condition checking whether the rate for the interval has already been exceeded:

```python
min_msg_rate = float(self._iopub_msg_count) / self.limit_poll_interval
min_data_rate = float(self._iopub_data_count) / self.limit_poll_interval

if (
    elapsed_time >= self.limit_poll_interval
    or min_data_rate >= self.iopub_data_rate
    or min_msg_rate >= self.iopub_msg_rate
):
```

In fact, maybe elapsed_time shouldn't be considered at all, just considering the total for a given window on every message.